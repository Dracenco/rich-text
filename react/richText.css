.container {}

.strong {}

.italic {}

.heading {}

.link {}

.paragraph {}

.image {}