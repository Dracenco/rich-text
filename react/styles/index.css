.container {}

.strong {}

.italic {}

.heading {}

.link {}

.paragraph {}

.image {}

.table {}

.tableHead {}

.tableBody {}
.heading-level-1 {}
.heading-level-2 {}
.heading-level-3 {}
.heading-level-4 {}
.heading-level-5 {}
.heading-level-6 {}

.list {}
.listItem {}
