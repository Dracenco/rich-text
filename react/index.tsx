import React, { FunctionComponent, memo, useEffect, useState } from 'react'
import marked, { Renderer } from 'marked'
import { values } from 'ramda'
import escapeHtml from 'escape-html'
import insane from 'insane'

 //@ts-ignore
import styles from './richText.css'

import {
  textPositionTypes,
  textAlignmentTypes,
  textPositionValues,
  textAlignmentValues,
} from './SchemaTypes'

const justifyTokens = {
  [textPositionValues.LEFT]: 'justify-start',
  [textPositionValues.CENTER]: 'justify-center',
  [textPositionValues.RIGHT]: 'justify-end',
}

const alignTokens = {
  [textAlignmentValues.LEFT]: 'tl',
  [textAlignmentValues.CENTER]: 'tc',
  [textAlignmentValues.RIGHT]: 'tr',
}

const itemsTokens = {
  [textAlignmentValues.LEFT]: 'items-start',
  [textAlignmentValues.CENTER]: 'items-center',
  [textAlignmentValues.RIGHT]: 'items-end',
}

type PropTokensNames = 'textPosition' | 'textAlignment'

const defaultValues = {
  textPosition: textPositionTypes.TEXT_POSITION_LEFT.value,
  textAlignment: textAlignmentTypes.TEXT_ALIGNMENT_LEFT.value,
}

const getEnumValues = (
  enumObject: Record<string, { value: string; name: string }>
) => values(enumObject).map(({ value }) => value)
const getEnumNames = (
  enumObject: Record<string, { value: string; name: string }>
) => values(enumObject).map(({ name }) => name)

const safelyGetToken = (
  tokenMap: Record<string, string>,
  valueWanted: string,
  propName: PropTokensNames
) => tokenMap[valueWanted] || defaultValues[propName]

interface Props {
  font: string
  text: string
  textAlignment: textAlignmentValues
  textColor: string
  textPosition: textPositionValues
}

interface VTEXIOComponent extends FunctionComponent<Props> {
  schema?: any
}

const sanitizerConfig = { 
  allowedTags: ['p', 'span', 'a', 'div', 'br', 'img'], 
  allowedAttributes:{
    a: ['class', 'href', 'title'],
    span: ['class'],
    p: ['class'],
    div: ['class'],
    img: ['class', 'src', 'title', 'alt'],
  }, 
}

const RichText: FunctionComponent<Props> = ({
  font,
  text,
  textAlignment,
  textColor,
  textPosition,
}) => {
  const [isMounted, setMounted] = useState(false)
  useEffect(() => {
    setMounted(true)
  }, [])

  if (!isMounted) {
    const renderer = new Renderer()
    renderer.paragraph = text => `<p class="lh-copy ${styles.paragraph}">${text}</p>`
    renderer.strong = text => `<span class="b ${styles.strong}">${text}</span>`
    renderer.em = text => `<span class="i ${styles.italic}">${text}</span>`
    renderer.heading = text => `<span class="${styles.heading}">${text}</span>`
    renderer.link = (href: string, title: string, text: string) =>
      `<a class="${styles.link}" href="${href}" ${
        title ? `title="${title}"` : ''
      }>${text}</a>`
    renderer.html = html => escapeHtml(html)
    renderer.image = (href: string, title: string, text: string) =>
      `<img class="${styles.image}" src="${href}" alt="${text}" title="${title}"/>`

    marked.setOptions({
      gfm: true,
      breaks: true,
      sanitize: false, //Use DOMPurify for sanitizing
      smartLists: true,
      renderer,
    })
  }

  const alignToken = safelyGetToken(alignTokens, textAlignment, 'textAlignment')
  const itemsToken = safelyGetToken(itemsTokens, textPosition, 'textPosition')
  const justifyToken = safelyGetToken(
    justifyTokens,
    textPosition,
    'textPosition'
  )

  const html = insane(marked(text), sanitizerConfig)
  return (
    <div
      className={`${
        styles.container
      } flex ${alignToken} ${itemsToken} ${justifyToken} ${font} ${textColor}`}
    >
      <div dangerouslySetInnerHTML={{ __html: html }} />
    </div>
  )
}

RichText.defaultProps = {
  font: 't-body',
  text: '',
  textPosition: defaultValues.textPosition,
  textAlignment: defaultValues.textAlignment,
  textColor: 'c-on-base',
}

const MemoizedRichText: VTEXIOComponent = memo(RichText)

MemoizedRichText.schema = {
  title: 'editor.rich-text.title',
  description: 'editor.rich-text.description',
  type: 'object',
  properties: {
    text: {
      title: 'editor.rich-text.text.title',
      description: 'editor.rich-text.textPosition.description',
      type: 'string',
      default: '',
    },
    textPosition: {
      title: 'editor.rich-text.textPosition.title',
      description: 'editor.rich-text.textPosition.description',
      type: 'string',
      enum: getEnumValues(textPositionTypes),
      enumNames: getEnumNames(textPositionTypes),
      default: defaultValues.textPosition,
    },
    textAlignment: {
      title: 'editor.rich-text.textAlignment.title',
      description: 'editor.rich-text.textAlignment.description',
      type: 'string',
      default: defaultValues.textAlignment,
      enum: getEnumValues(textAlignmentTypes),
      enumNames: getEnumNames(textAlignmentTypes),
    },
    font: {
      title: 'editor.rich-text.font.title',
      description: 'editor.rich-text.font.description',
      type: 'string',
      default: 't-body',
    },
    textColor: {
      title: 'editor.rich-text.textColor.title',
      description: 'editor.rich-text.textColor.description',
      type: 'string',
      default: 'c-on-base',
    },
  },
}

export default MemoizedRichText
