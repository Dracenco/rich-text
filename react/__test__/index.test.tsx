/* eslint-env jest */
import React from 'react'
import { render } from '@vtex/test-tools/react'
import RichText from '../index'

import { textPositionValues, textAlignmentValues } from '../SchemaTypes'

describe('Rich text component', () => {
  const defaultProps = {
    font: 't-body',
    text: '',
    textPosition: textPositionValues.LEFT,
    textAlignment: textAlignmentValues.LEFT,
    textColor: 'c-on-base',
  }
  it('should render with bold', () => {
    const component = render(
      <RichText {...defaultProps} text={'**be bold**'} />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render with paragraph', () => {
    const component = render(
      <RichText {...defaultProps} text={'IAM BOLD'} />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render link with title', () => {
    const component = render(
      <RichText
        {...defaultProps}
        text={
          '[I\'m an inline-style link with title](https://www.google.com "Google\'s Homepage")\n**Bollllddd**\n*this is talic*'
        }
      />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render and sanitize malicious text', () => {
    const component = render(
      <RichText
        {...defaultProps}
        text={
          "Hi everyone! This is **how** you do XSS <script>alert('XSS')</script>"
        }
      />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render and sanitize malicious text', () => {
    const component = render(
      <RichText {...defaultProps} text={'[some text](javascript:exec())'} />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render and sanitize malicious javascript', () => {
    const component = render(
      <RichText
        {...defaultProps}
        text={'<a href="javascript:exec()">HEHEHE</a>'}
      />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render and sanitize malicious javascript', () => {
    const component = render(
      <RichText
        {...defaultProps}
        text={
          '<DIV STYLE="background-image: url(&#1;javascript:alert(\'XSS\'))">'
        }
      />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render with image', () => {
    const component = render(
      <RichText
        {...defaultProps}
        text={
          `Inline-style: 
          ![alt text](https://github.com/adam-p/markdown-here/raw/master/src/common/images/icon48.png "Logo Title Text 1")`
        }
      />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render with image block class props', () => {
    const component = render(
      <RichText
        {...defaultProps}
        text={
          `[I\'m an inline-style link with title](https://www.google.com "Google\'s Homepage")\n**Bollllddd**\n*this is talic*`
        }
        blockClass="home"
      />
    )
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
})

