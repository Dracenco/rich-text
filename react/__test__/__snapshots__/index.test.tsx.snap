// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rich text component should render an iframe 1`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        <span
          class="b strong"
        >
          <iframe
            frameborder="0"
            height="100%"
            src="https://invictastores.typeform.com/to/KmycOT?typeform-embed=embed-widget&embed-hide-footer=true&embed-hide-headers=true&embed-opacity=50&typeform-embed-id=d0pe3"
            style="border: 0px;"
            width="100%"
          />
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render and sanitize malicious javascript 1`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        <a>
          HEHEHE
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render and sanitize malicious javascript 2`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      &lt;DIV STYLE="background-image: url(&#1;javascript:alert('XSS'))"&gt;
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render and sanitize malicious text 1`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        Hi everyone! This is 
        <span
          class="b strong"
        >
          how
        </span>
         you do XSS 
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render and sanitize malicious text 2`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        <a
          class="link"
        >
          some text
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render link with title 1`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        <a
          class="link"
          href="https://www.google.com"
          title="Google's Homepage"
        >
          I'm an inline-style link with title
        </a>
        <br />
        <span
          class="b strong"
        >
          Bollllddd
        </span>
        <br />
        <span
          class="i italic"
        >
          this is talic
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render with bold 1`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        <span
          class="b strong"
        >
          be bold
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render with image 1`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        Inline-style:
        <br />
                  
        <img
          alt="alt text"
          class="image"
          src="https://github.com/adam-p/markdown-here/raw/master/src/common/images/icon48.png"
          title="Logo Title Text 1"
        />
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render with image block class props 1`] = `
<DocumentFragment>
  <div
    class="container container--home flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        <a
          class="link"
          href="https://www.google.com"
          title="Google's Homepage"
        >
          I'm an inline-style link with title
        </a>
        <br />
        <span
          class="b strong"
        >
          Bollllddd
        </span>
        <br />
        <span
          class="i italic"
        >
          this is talic
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rich text component should render with paragraph 1`] = `
<DocumentFragment>
  <div
    class="container flex tl items-start justify-start t-body c-on-base"
  >
    <div>
      <p
        class="lh-copy paragraph"
      >
        IAM BOLD
      </p>
    </div>
  </div>
</DocumentFragment>
`;
